File: src/main/java/queue/QueueLinkedList.java
package queue;
/**
 * A simple implementation of a Queue using a linked list.
 */
public class QueueLinkedList<T> {

    /**
     * Inner class representing a node in the linked list.
     */
    private static class Node<T> {
        private T data; // The value stored in the node
        private Node<T> next; // Pointer to the next node

        public Node(T data) {
            this.data = data;
            this.next = null;
        }
    }
    private Node<T> front; // Points to the front of the queue
    private Node<T> rear;  // Points to the rear of the queue
    private int size;      // Tracks the size of the queue

    /**
     * Constructor to initialize an empty queue.
     */
    public QueueLinkedList() {
        front = null;
        rear = null;
        size = 0;
    }

     /**
     * Add an element to the rear of the queue.
     * @param value The value to be added.
     */
    public void enqueue(T value) {
        Node<T> newNode = new Node<>(value);
        if (isEmpty()) { //if rear = null, if queue is empty
            f front = rear = newNode;
        } else {
            rear.next = newNode; // Link the new node at the rear
            rear = newNode;      // Update the rear pointer
        }
        size++;
    }
    
    /**
     * Remove and return the element from the front of the queue.
     * @return The value of the removed element, or null if the queue is empty.
     */
    public T dequeue() {
        if (isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        }
        T value = front.data; // Get the value from the front node
        front = front.next;   // Move the front pointer to the next node
        if (front == null) {  // If the queue becomes empty
            rear = null;      // Reset the rear pointer
        }
        size--;
        return value;
    }
    
    /**
     * View the element at the front of the queue without removing it.
     * @return The value at the front of the queue, or null if the queue is empty.
     */
    public T peek() {
        if (isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        }
        return front.data;
    }
    

       /**
     * Check if the queue is empty.
     * @return True if the queue is empty, false otherwise.
     */
    public boolean isEmpty() {
        return size == 0;
    }
    

    /**
     * Get the current size of the queue.
     * @return The number of elements in the queue.
     */
    public int size() {
        return size;
    }
    
    public void clear() {
        front = rear = null;
        size = 0;
    }

     /**
     * Print all elements in the queue from front to rear.
     */
    public void printQueue() {
        if (isEmpty()) {
            System.out.println("Queue is empty!");
            return;
        }
        Node<T> current = front;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }
}

File: python/queue/QueueLinkedList.py
class ListNode:
    """
    Class representing a single node in the linked list.
    """
    def __init__(self, value):
        self.value = value  # Value of the node
        self.next = None    # Pointer to the next node


"""
Queue implemented using a linked list
- Enqueue: (last node) O(1)
- Dequeue: (first node) O(1)
- Peek: (first node) O(1)


"""


class QueueLinkedList:
    """
    Queue implementation using a linked list.
    """
    def __init__(self):
        """
        Initialize the queue with pointers to the front and rear.
        """
        self.front = None  # Points to the front of the queue
        self.rear = None   # Points to the rear of the queue
        self.size = 0      # Tracks the size of the queue

    def enqueue(self, value):
        """
        Add an element to the rear of the queue.
        :param value: The value to add.
        """
        new_node = ListNode(value)  # Create a new node
        if not self.rear:  # If the queue is empty
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node  # Link the new node to the rear
            self.rear = new_node       # Update the rear pointer
        self.size += 1

    def dequeue(self):
        """
        Remove and return the front element of the queue.
        :return: The front element, or None if the queue is empty.
        """
        if not self.front:  # If the queue is empty
            print("Queue is empty!")
            return None
        value = self.front.value  # Get the value of the front node
        self.front = self.front.next  # Move the front pointer to the next node
        if not self.front:  # If the queue becomes empty
            self.rear = None  # Update the rear pointer
        self.size -= 1
        return value

    def peek(self):
        """
        View the front element of the queue without removing it.
        :return: The front element, or None if the queue is empty.
        """
        if not self.front:
            print("Queue is empty!")
            return None
        return self.front.value

    def isEmpty(self):
        """
        Check if the queue is empty.
        :return: True if empty, False otherwise.
        """
        return self.front is None

    def getSize(self):
        """
        Get the current size of the queue.
        :return: The size of the queue.
        """
        return self.size

    def printQueue(self):
        """
        Print all elements in the queue from front to rear.
        """
        current = self.front
        while current:
            print(current.value, end=" -> ")
            current = current.next
        print() #new line

